Patch-Source: https://github.com/aufflick/p5-http-daemon-ssl/commit/8b252a5a5e54249c45440b9e21e4db03888ab7d3

but also edited by ptrcnull to fix the certificates and test number and a few other things ugh

From 8b252a5a5e54249c45440b9e21e4db03888ab7d3 Mon Sep 17 00:00:00 2001
From: Mark Aufflick <mark@aufflick.com>
Date: Sun, 28 Sep 2008 10:53:34 +0000
Subject: [PATCH] improved test from Steffen Ullrich

---
 t/testmodule.t | 47 +++++++++++++++++++++++++----------------------
 1 file changed, 25 insertions(+), 22 deletions(-)

diff --git a/certs/server-cert.pem b/certs/server-cert.pem
index 0fc5c24..3dbcb9a 100644
--- a/certs/server-cert.pem
+++ b/certs/server-cert.pem
@@ -1,44 +1,21 @@
-Certificate:
-    Data:
-        Version: 1 (0x0)
-        Serial Number: 2 (0x2)
-        Signature Algorithm: md5WithRSAEncryption
-        Issuer: C=US, ST=Some-State, O=Dummy IO::Socket::SSL Certificate Authority, CN=Dummy IO::Socket::SSL Certificate Authority
-        Validity
-            Not Before: Jul 20 16:06:37 2002 GMT
-            Not After : Dec  5 16:06:37 2029 GMT
-        Subject: C=US, ST=Some-State, O=IO::Socket::SSL Dummy Server Certificate, CN=IO::Socket::SSL Dummy Server Certificate
-        Subject Public Key Info:
-            Public Key Algorithm: rsaEncryption
-            RSA Public Key: (512 bit)
-                Modulus (512 bit):
-                    00:9f:27:5f:4a:8a:35:4a:7f:3f:d1:80:25:96:26:
-                    0a:da:af:9a:6d:bc:23:ba:71:91:5b:40:d1:2d:2b:
-                    c8:60:2a:ef:e9:54:e5:a2:64:0a:57:90:35:bf:cd:
-                    b6:36:f3:25:53:68:65:2c:d8:d0:f9:b7:f3:7f:2e:
-                    f8:e2:3d:e0:dd
-                Exponent: 65537 (0x10001)
-    Signature Algorithm: md5WithRSAEncryption
-        57:a7:2d:91:cc:e9:11:16:bb:c1:cd:b5:a5:e1:26:99:8f:ee:
-        8c:b0:2d:b6:54:f4:8a:8e:fd:8f:45:9a:68:d8:0e:ef:d6:a5:
-        38:6a:48:d0:08:da:a8:87:3c:70:05:18:69:a1:c8:ee:94:a7:
-        87:40:f5:4f:64:b4:b0:c6:d3:d2:ed:f9:cc:d1:fe:da:4d:99:
-        4d:22:02:f6:0e:9b:c0:cc:42:59:50:2f:5c:fc:5b:70:f9:0b:
-        ec:6e:5b:eb:d7:6f:a1:b8:67:57:b1:4f:99:bd:ad:03:9d:b5:
-        f3:44:5c:36:1c:fa:33:82:87:0b:99:aa:f5:39:5c:63:23:6b:
-        48:2d
 -----BEGIN CERTIFICATE-----
-MIICQzCCAawCAQIwDQYJKoZIhvcNAQEEBQAwgY4xCzAJBgNVBAYTAlVTMRMwEQYD
-VQQIEwpTb21lLVN0YXRlMTQwMgYDVQQKEytEdW1teSBJTzo6U29ja2V0OjpTU0wg
-Q2VydGlmaWNhdGUgQXV0aG9yaXR5MTQwMgYDVQQDEytEdW1teSBJTzo6U29ja2V0
-OjpTU0wgQ2VydGlmaWNhdGUgQXV0aG9yaXR5MB4XDTAyMDcyMDE2MDYzN1oXDTI5
-MTIwNTE2MDYzN1owgYgxCzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpTb21lLVN0YXRl
-MTEwLwYDVQQKEyhJTzo6U29ja2V0OjpTU0wgRHVtbXkgU2VydmVyIENlcnRpZmlj
-YXRlMTEwLwYDVQQDEyhJTzo6U29ja2V0OjpTU0wgRHVtbXkgU2VydmVyIENlcnRp
-ZmljYXRlMFwwDQYJKoZIhvcNAQEBBQADSwAwSAJBAJ8nX0qKNUp/P9GAJZYmCtqv
-mm28I7pxkVtA0S0ryGAq7+lU5aJkCleQNb/NtjbzJVNoZSzY0Pm3838u+OI94N0C
-AwEAATANBgkqhkiG9w0BAQQFAAOBgQBXpy2RzOkRFrvBzbWl4SaZj+6MsC22VPSK
-jv2PRZpo2A7v1qU4akjQCNqohzxwBRhpocjulKeHQPVPZLSwxtPS7fnM0f7aTZlN
-IgL2DpvAzEJZUC9c/Ftw+Qvsblvr12+huGdXsU+Zva0DnbXzRFw2HPozgocLmar1
-OVxjI2tILQ==
+MIIDYTCCAkmgAwIBAgIFANkVla8wDQYJKoZIhvcNAQELBQAwIjEgMB4GA1UEAwwX
+SU86OlNvY2tldDo6U1NMIERlbW8gQ0EwHhcNMjIxMjExMTk1MzQxWhcNMzIxMjA4
+MTk1MzQxWjAXMRUwEwYDVQQDDAxzZXJ2ZXIubG9jYWwwggEiMA0GCSqGSIb3DQEB
+AQUAA4IBDwAwggEKAoIBAQCnMcTzSybDMjCCFTfPPzOltpavJ1cvOQ4X99q7jQph
+2dTGx1feefwcuKJl3eEuwiV/y6MWWkjJVC1vICSu2BuBhL76jCgl0mKIQbN3jVpS
+KtqnytRGVvGvB3AP71RMzRXaI0xiwRsjvXnBhliTaYBtbpVqry1Cx7eouxeveRxx
+3+5dfBNU0i9U18EZPl99Yl2z2Z6OvzT0ULJl9cWP90UKrX16G5eH8vHrMwm02rpn
+i+7u0o7O9a7/xQV28cSoEgp2Cnbg0ZUXbmQS4aYDqIkpS2GlOL8eV26KvM2hYX7h
+qy0CsrjJ4riJd+YhmGRsPH3DBGjB/kRX8NhAP2+tblc/AgMBAAGjgagwgaUwHQYD
+VR0OBBYEFHW7Ml+/HDstKVpiCxHde7b+VttWMB8GA1UdIwQYMBaAFEnT2LwqEtZv
+wVkEbtlv/7SmEt9cMB0GA1UdEQQWMBSCDHNlcnZlci5sb2NhbIcEfwAAATAMBgNV
+HRMBAf8EAjAAMA4GA1UdDwEB/wQEAwIFoDARBglghkgBhvhCAQEEBAMCBkAwEwYD
+VR0lBAwwCgYIKwYBBQUHAwEwDQYJKoZIhvcNAQELBQADggEBAKA5/2fl2oRtBnUj
+Zr+a2Z1uc+oTP03VPT/w46uolz27MqgQyBiSX+a2WWFWZJZFDK6jv3Yd1C7j+KOm
+V7sbHOhoIGDwQC55vwdlc5r72RYZOuZSFtujvaABEZ+vF8AHnI3PbiShedL/bK2N
+yZYWtBj4Lbl1Hb9I+AjOY5TJ5zcenyS5hIEYXZgV0NH5Thf4zMIKrRZ6//3XcN5n
+zT7nMyPTqh0nYIAblmOKvYu6RJQ29BL8FyNmNXjItr3HjaKIxZry7apvwrHBt+a7
+bLQzc5e8/cb06gTHZJYdsWDBT6Mv81jNFA/d2OEbpWCNH4ySLPHCBItMmWTxZR87
+D7hgP1A=
 -----END CERTIFICATE-----
diff --git a/certs/server-key.pem b/certs/server-key.pem
index b7a165f..cf08a7d 100644
--- a/certs/server-key.pem
+++ b/certs/server-key.pem
@@ -1,9 +1,28 @@
------BEGIN RSA PRIVATE KEY-----
-MIIBPAIBAAJBAJ8nX0qKNUp/P9GAJZYmCtqvmm28I7pxkVtA0S0ryGAq7+lU5aJk
-CleQNb/NtjbzJVNoZSzY0Pm3838u+OI94N0CAwEAAQJAf/DavcVVCco5t2TY0ldK
-qno4Hrb70cmyHDWC8lkb/5HAGbCGxpsstXxVKczRO201vcFUKm6PX5moUnFCINpg
-UQIhAM+ooHbD0eLL0K6limEnW7GId/+DFI/6KFXk2Nzm//XXAiEAxDQbWQvZS8DO
-HJ5JV8flvMhH30KLeH+zpsvBjWJK4GsCIQCUF7woNsquJZBznNctJjZ8S8jYThES
-BONTLluCXrNYDQIhAJFnsHDQqCxM6jMpV193pJnAsAsUbPpTYZeWX43hL26bAiEA
-jNB3PPNvTNr5tICkO/lMZcN87eUn4ZAtrNzCVF5ilEo=
------END RSA PRIVATE KEY-----
+-----BEGIN PRIVATE KEY-----
+MIIEvAIBADANBgkqhkiG9w0BAQEFAASCBKYwggSiAgEAAoIBAQCnMcTzSybDMjCC
+FTfPPzOltpavJ1cvOQ4X99q7jQph2dTGx1feefwcuKJl3eEuwiV/y6MWWkjJVC1v
+ICSu2BuBhL76jCgl0mKIQbN3jVpSKtqnytRGVvGvB3AP71RMzRXaI0xiwRsjvXnB
+hliTaYBtbpVqry1Cx7eouxeveRxx3+5dfBNU0i9U18EZPl99Yl2z2Z6OvzT0ULJl
+9cWP90UKrX16G5eH8vHrMwm02rpni+7u0o7O9a7/xQV28cSoEgp2Cnbg0ZUXbmQS
+4aYDqIkpS2GlOL8eV26KvM2hYX7hqy0CsrjJ4riJd+YhmGRsPH3DBGjB/kRX8NhA
+P2+tblc/AgMBAAECggEABJVkCtodPpivpRaj0wtZL8p+UwrxuZpc0oy5nblTdt9G
+lV8oVNxklvz7fBjFxZxjnsnxt05+VFakcDl3XVEQtU+dqgy8RQfW1QQdbBefSZq3
+J9vIT1gELteLW5nPZn5GLRbD+f5v7147FPJz7Ial6K9xaof8O6px/y7cirOinf80
+Ll73KxTyb7amgAxJS34/STSHvBGUu0RYUQWX7cXllqONn+zZ+fgiertwervHYH+7
+rkcbAsG3AGZtXJ20K20qOmc5QvtIdu0OGvRdW861ZYCNgEcUaeO7Lvt9CrOZyhUe
+lqGw22cxJevIPUEoJY4gyNY3SV/WmqG+QKOIK4IuMQKBgQDoqpVvkp3Q47GsG6Dr
+skTgIv9Aof7/4fv9dHNYWYaUzQMGW2uxr7Dy6yuvhkplwjVTYxmZrwcKeT0L4wuu
+ofhSPRKH7h4o3CVZI9QSz6hrk15u9oKvqmN9W5FOj5ZaXxcdT1NHVENGMYl08E2J
+dzLgTJPFPnCEWiZKEE0QSLW4LQKBgQC39kalXeDAMDXR7Db0ui1Wgfc0hL0NuDim
+HrzmgtZrZCoYZLjvm1pYQ0sxZ/8S97oh3HKTZh83plbmDQzTRjgSJTGLp0utRTuY
+2TuyJURurX2SJggg+6yL0o2eS1yA4t1Mb2onr49o4DSeEggRML9hAY3Ihg2cTYiy
+ImTQ8vekmwKBgC+4nUHvLpNjwFNur0jonZvjUbtt/qF5Nng75FSguCvZCN/K7IHb
+aU3J0oID50qL1OgvkVamQalySIUhoonFCuvDPwPGYUU8MiTgZmUdVowKA/p6cT+a
+kSFrIJiedtY+Xr1SQeCFde71xh3IE/84BaVfz4dLUUS0QNo8EbJfV3ZZAoGACbwS
+iPWqywDCGFWzosenVoiSGEld57fz53aA8IHD7vLh92B9GNDTuw/0jqy+JrbNNrV/
+qqUgycUXnBzcrOFuXidxs74qlwSu3qvAKPEn6eNsXat9iqFGxC9kJxg90OQwabcL
+mwYDRL14i1TQ8Hfv6KY4ZoARgE/qB+MiCpyQ1jkCgYAD3jMZAYaxp11Zl0qxffCT
+AQZkah+tTA8tC0TYSNxUUq18nnU8gvLuIF8YUt/HJkFajA9GQkA0rg+KUZ3ig3n2
+VfwHCMf0HGH90jc9wRQRd0FlkaAn68e5t3/eCAoQFnN65iit+ODR8isqTqRISMJt
+nL6o91SHe9luE7bU49fnVQ==
+-----END PRIVATE KEY-----
diff --git a/certs/test-ca.pem b/certs/test-ca.pem
index 36bf8e4..2b5e7fa 100644
--- a/certs/test-ca.pem
+++ b/certs/test-ca.pem
@@ -1,21 +1,19 @@
 -----BEGIN CERTIFICATE-----
-MIIDgzCCAuygAwIBAgIBADANBgkqhkiG9w0BAQQFADCBjjELMAkGA1UEBhMCVVMx
-EzARBgNVBAgTClNvbWUtU3RhdGUxNDAyBgNVBAoTK0R1bW15IElPOjpTb2NrZXQ6
-OlNTTCBDZXJ0aWZpY2F0ZSBBdXRob3JpdHkxNDAyBgNVBAMTK0R1bW15IElPOjpT
-b2NrZXQ6OlNTTCBDZXJ0aWZpY2F0ZSBBdXRob3JpdHkwHhcNMDIwNzIwMTYwNTU0
-WhcNMjkxMjA1MTYwNTU0WjCBjjELMAkGA1UEBhMCVVMxEzARBgNVBAgTClNvbWUt
-U3RhdGUxNDAyBgNVBAoTK0R1bW15IElPOjpTb2NrZXQ6OlNTTCBDZXJ0aWZpY2F0
-ZSBBdXRob3JpdHkxNDAyBgNVBAMTK0R1bW15IElPOjpTb2NrZXQ6OlNTTCBDZXJ0
-aWZpY2F0ZSBBdXRob3JpdHkwgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBALQm
-bgkEUWImNkjWcO6qn5NZ7rCFbtrzqEYbqciy+1qlWuoBgU44n9ykD1c/BcmBPsDT
-bIOfLzjcdJj38taXu7kcRclchJ+/c6o/SmDv7UqcL6QgVSZRvRrK7TDypMqe3sW8
-zCvTF8WtSsgFy5f9qlUdx4NowMzVV7OFl+6x4YlpAgMBAAGjge4wgeswHQYDVR0O
-BBYEFDU4SrHVMHDjd2kBgFM/qyC3DPxFMIG7BgNVHSMEgbMwgbCAFDU4SrHVMHDj
-d2kBgFM/qyC3DPxFoYGUpIGRMIGOMQswCQYDVQQGEwJVUzETMBEGA1UECBMKU29t
-ZS1TdGF0ZTE0MDIGA1UEChMrRHVtbXkgSU86OlNvY2tldDo6U1NMIENlcnRpZmlj
-YXRlIEF1dGhvcml0eTE0MDIGA1UEAxMrRHVtbXkgSU86OlNvY2tldDo6U1NMIENl
-cnRpZmljYXRlIEF1dGhvcml0eYIBADAMBgNVHRMEBTADAQH/MA0GCSqGSIb3DQEB
-BAUAA4GBAIbCsK/qUXiIsRvg1ptaLNM6VsuR8ifNrmo9A4zk1h4OCixys6Hmoow6
-3MndnLpD3rh3UCYh0M20+fiHcwSmHZvBo3dfSSvYnH0gFSBjKp/wgGcb3Cvl3dRX
-aeWZGrKQKLI6DrHqAiSu9rv+2kfzgmRLt0K+gdb2GkQqCBwT8Gjr
+MIIDIjCCAgqgAwIBAgIEPGupAjANBgkqhkiG9w0BAQsFADAiMSAwHgYDVQQDDBdJ
+Tzo6U29ja2V0OjpTU0wgRGVtbyBDQTAeFw0yMjEyMTExOTUzNDFaFw0zMjEyMDgx
+OTUzNDFaMCIxIDAeBgNVBAMMF0lPOjpTb2NrZXQ6OlNTTCBEZW1vIENBMIIBIjAN
+BgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAunYRHBif+Nltt/y4zxgrmg0/XTMh
+CVj1+ITPtpIImmuNlN/YKFzjdVLCjsfsiy7h1a8IfqopzFp/JciijHXEW9NyqWJY
+hwGATku0W72e2T4R0H0LikJ33R1oTP5YfKOHz82ZTJJF3xI57py9Q6AJjM1lCMt6
+1EAUqPteBLy0xIOabQDp+RpzKv1Uo5aMHiLACl6J52w8IBDIYhz5YXDNULdt6AYK
+9Hqyw+m6FiULr+SrKfjgHFHunM9r79aS27iuqwLrAiHvuCfSp0o+qmslz6YIfpF6
+/BziErz1dtnDjqinUlGYCDUPRb2nHI5/h6uFGQuF+w3hud3Ip/nZLMLPtQIDAQAB
+o2AwXjAdBgNVHQ4EFgQUSdPYvCoS1m/BWQRu2W//tKYS31wwCQYDVR0jBAIwADAP
+BgNVHRMBAf8EBTADAQH/MA4GA1UdDwEB/wQEAwIChDARBglghkgBhvhCAQEEBAMC
+AAcwDQYJKoZIhvcNAQELBQADggEBAI1f2QhRUn4pnrESAYowXBPFe0SKpN8d2Pqz
+WbagBA2/QaZmkW7AhFoSb0Ni2Jhe9RflbBfF6WQIS6uVuSWT414DGMUQjhGjhLXm
+5f4YxtIpeKyFqAkcpHiB7XEO+mfT1iV5jw1e4yq7c8fOFCEI4cAQzdDpr35vjEPw
+pdSXNHCqIbmbDRZigwGK3m0WNATu2K+54QC+rPKzzb6CDb24lRcWKlmzbaeowKEn
+zP5GqlNz9PTF6OTxBFxe6hxklE6uudViEqBkD3TvFeXDt4ny1xFUbABxtU3qbwHA
+r//e1FKMu3RDxJyAiTVPKwsN8dL6KgH4AFL96HV/pkCrIds7i3c=
 -----END CERTIFICATE-----
diff --git a/t/testmodule.t b/t/testmodule.t
index d03aeef..5612b41 100644
--- a/t/testmodule.t
+++ b/t/testmodule.t
@@ -20,10 +20,37 @@ foreach ($^O) {
 
 print "1..$numtests\n";
 
-$test = 0;
+$test = 1;
+
+my $server = new HTTP::Daemon::SSL(
+				   LocalAddr => '127.0.0.1',
+				   LocalPort => 0,
+				   Listen => 5,
+				   Timeout => 30,
+				   ReuseAddr => 1,
+				   SSL_server => 1,
+				   SSL_verify_mode => 0x00,
+				   SSL_ca_file => "certs/test-ca.pem",
+				   SSL_cert_file => "certs/server-cert.pem",
+				   SSL_key_file => "certs/server-key.pem");
+
+if (!$server) {
+    print "not ok $test\n";
+    exit;
+}
+$SSL_SERVER_PORT = $server->sockport;
+&ok("server init port=$SSL_SERVER_PORT");
+
+
+print "not " if (!defined fileno($server));
+&ok("server fileno");
+
+print "not " unless ($server->url =~ m!^https:!);
+&ok("server url test");
+
 
 unless (fork) {
-    sleep 1;
+    close($server);
 
     my $client = new IO::Socket::INET(PeerAddr => $SSL_SERVER_ADDR,
 				      PeerPort => $SSL_SERVER_PORT);
@@ -52,27 +79,6 @@ unless (fork) {
 }
 
 
-my $server = new HTTP::Daemon::SSL(LocalPort => $SSL_SERVER_PORT,
-				   LocalAddr => $SSL_SERVER_ADDR,
-				   Listen => 5,
-				   Timeout => 30,
-				   ReuseAddr => 1,
-				   SSL_verify_mode => 0x00,
-				   SSL_ca_file => "certs/test-ca.pem",
-				   SSL_cert_file => "certs/server-cert.pem");
-
-if (!$server) {
-    print "not ok $test\n";
-    exit;
-}
-&ok("server init");
-
-print "not " if (!defined fileno($server));
-&ok("server fileno");
-
-print "not " unless ($server->url =~ m!^https:!);
-&ok("server url test");
-
 my $conn;
 if (!($conn = $server->accept)) {
     # first client request is a bad request
